src/reflect/scala/reflect/internal/Definitions.scala

```diff
@@ -813,16 +843,27 @@ trait Definitions extends api.StandardDefinitions {
      * The class defining the method is a supertype of `tp` that
      * has a public no-arg primary constructor.
      */
-    def samOf(tp: Type): Symbol = if (!settings.Xexperimental) NoSymbol else findSam(tp)
-
-    def findSam(tp: Type): Symbol = {
-      // if tp has a constructor, it must be public and must not take any arguments
-      // (not even an implicit argument list -- to keep it simple for now)
-      val tpSym  = tp.typeSymbol
-      val ctor   = tpSym.primaryConstructor
-      val ctorOk = !ctor.exists || (!ctor.isOverloaded && ctor.isPublic && ctor.info.params.isEmpty && ctor.info.paramSectionCount <= 1)
+    def samOf(tp: Type): Symbol = {
+      // look at erased type because we (only) care about what ends up in bytecode
+      // (e.g., an alias type or intersection type is fine as long as the intersection dominator compiles to an interface)
+      val tpSym = erasure.javaErasure(tp).typeSymbol
+
+      if (tpSym.exists
+          // We use Java's MetaLambdaFactory, which requires an interface for the sam's owner
+          // (TODO: Can't use isInterface, yet, as it hasn't been updated for the new trait encoding)
+          && (tpSym.isJavaInterface || tpSym.isTrait)
```

Comment:
> How about making this platform specific and being more liberal when targeting JS?

I don't think that's a good idea. One of the key strengths of Scala.js is that most Scala code compiles both on the JVM and on JS. Introducing platform-specific typechecker behavior would go against this powerful property.

Something that could make sense would be to always turn it on for `js.FunctionN` and `js.ThisFunctionN`, but not for other non-JVM-SAM types. Since `js.Function`s are JS-specific, they can only appear in non-cross-compiling code, and that would be fine. But that means specializing these specific types in the normal typechecker, which is awkward. A better way to do it would be to invent an annotation `@SAMIntended` (better name TBD) to be placed on such types that are not JVM SAMs but that are intended for the SAM treatment. We would happily put that annotation on all `js.Function` types. And it could also be beneficial for other classes/traits in JVM-specific and cross-compiling code.

I don't have a strong opinion between (a) all (non-JVM) SAMs receive the SAM treatment and (b) only `@SAMIntended` types receive the SAM treatment. I'm also open to other alternatives; these are only my suggestions.